File: java/nitf/src/java/nitf/SubWindow.java
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 * 
 * (C) Copyright 2004 - 2008, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public 
 * License along with this program; if not, If not, 
 * see <http://www.gnu.org/licenses/>.
 *
 */

package nitf;

/**
 * An object that contains information for reading image data.
 */
public final class SubWindow extends DestructibleObject
{

    /**
     * Creates a new SubWindow object
     * 
     * @throws NITFException
     */
    public SubWindow() throws NITFException
    {
        construct();
    }

    /**
     * @see DestructibleObject#DestructibleObject(long)
     */
    SubWindow(long address)
    {
        super(address);
    }

    /**
     * Destructs the underlying memory
     */
    protected native synchronized void destructMemory();

    /**
     * Returns the BandList array
     * 
     * @return
     */
    public native synchronized int[] getBandList();

    /**
     * Constructs the underlying memory
     * 
     * @throws NITFException
     */
    private native synchronized void construct() throws NITFException;

    /**
     * Sets the number of columns
     * 
     * @param numCols
     */
    public native synchronized void setNumCols(int numCols);

    /**
     * Sets the BandList array
     * 
     * @param bandList
     */
    public native synchronized void setBandList(int[] bandList);

    /**
     * Returns the start row
     * 
     * @return
     */
    public native synchronized int getStartRow();

    /**
     * Sets the start row
     * 
     * @param startRow
     */
    public native synchronized void setStartRow(int startRow);

    /**
     * Returns the number of rows
     * 
     * @return
     */
    public native synchronized int getNumRows();

    /**
     * Sets the number of rows
     * 
     * @param numRows
     */
    public native synchronized void setNumRows(int numRows);

    /**
     * Returns the start column
     * 
     * @return
     */
    public native synchronized int getStartCol();

    /**
     * Sets the start column
     * 
     * @param startCol
     */
    public native synchronized void setStartCol(int startCol);

    /**
     * Returns the number of columns
     * 
     * @return
     */
    public native synchronized int getNumCols();

    /**
     * Returns the number of bands
     * 
     * @return
     */
    public native synchronized int getNumBands();

    /**
     * Sets the number of bands
     * 
     * @param numBands
     */
    public native synchronized void setNumBands(int numBands);

    /**
     * Sets the DownSampler for this SubWindow If none is specified, or if
     * downSampler is null, the data will not be downsampled
     * 
     * @param downSampler
     */
    public native synchronized void setDownSampler(DownSampler downSampler);

    /**
     * Returns the DownSampler for this window, or null if none is specified
     * 
     * @return
     */
    public native synchronized DownSampler getDownSampler();

    public String toString()
    {
        StringBuffer buf = new StringBuffer();
        buf.append(SubWindow.class.getName() + ":[");
        buf.append("startCol=" + getStartCol());
        buf.append(",startRow=" + getStartRow());
        buf.append(",numCols=" + getNumCols());
        buf.append(",numRows=" + getNumRows());
        buf.append(",numBands=" + getNumBands());
        buf.append("]");
        return buf.toString();
    }

}

File: modules/c++/nitf/source/SubWindow.cpp
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 *
 * (C) Copyright 2004 - 2010, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include "nitf/SubWindow.hpp"

using namespace nitf;

SubWindow::SubWindow(const SubWindow & x)
{
    setNative(x.getNative());
}

SubWindow & SubWindow::operator=(const SubWindow & x)
{
    if (&x != this)
        setNative(x.getNative());
    return *this;
}

SubWindow::SubWindow(nitf_SubWindow * x)
{
    setNative(x);
    getNativeOrThrow();
}

SubWindow::SubWindow() throw(nitf::NITFException) : mDownSampler(NULL)
{
    setNative(nitf_SubWindow_construct(&error));
    getNativeOrThrow();
    setManaged(false);
}

SubWindow::~SubWindow()
{
    if (isValid() && getNative()->downsampler)
    {
        nitf::DownSampler ds(getNativeOrThrow()->downsampler);
        //decrement the current DownSampler
        ds.decRef();
    }
}

nitf::Uint32 SubWindow::getStartRow() const
{
    return getNativeOrThrow()->startRow;
}

void SubWindow::setStartRow(nitf::Uint32 value)
{
    getNativeOrThrow()->startRow = value;
}

nitf::Uint32 SubWindow::getNumRows() const
{
    return getNativeOrThrow()->numRows;
}

void SubWindow::setNumRows(nitf::Uint32 value)
{
    getNativeOrThrow()->numRows = value;
}

nitf::Uint32 SubWindow::getStartCol() const
{
    return getNativeOrThrow()->startCol;
}

void SubWindow::setStartCol(nitf::Uint32 value)
{
    getNativeOrThrow()->startCol = value;
}

nitf::Uint32 SubWindow::getNumCols() const
{
    return getNativeOrThrow()->numCols;
}

void SubWindow::setNumCols(nitf::Uint32 value)
{
    getNativeOrThrow()->numCols = value;
}

nitf::Uint32 SubWindow::getBandList(int i)
{
    return getNativeOrThrow()->bandList[i];
}

void SubWindow::setBandList(nitf::Uint32 * value)
{
    getNativeOrThrow()->bandList = (nitf_Uint32*)value;
}

nitf::Uint32 SubWindow::getNumBands() const
{
    return getNativeOrThrow()->numBands;
}

void SubWindow::setNumBands(nitf::Uint32 value)
{
    getNativeOrThrow()->numBands = value;
}

void SubWindow::setDownSampler(nitf::DownSampler* downSampler)
    throw (nitf::NITFException)
{
    if (getNativeOrThrow()->downsampler)
    {
        nitf::DownSampler ds(getNativeOrThrow()->downsampler);
        //decrement the current DownSampler
        ds.decRef();
    }

    //increment the reference for this DownSampler
    getNativeOrThrow()->downsampler = downSampler->getNative();
    downSampler->incRef();
    mDownSampler = downSampler;
}


nitf::DownSampler* SubWindow::getDownSampler() throw (nitf::NITFException)
{
    return mDownSampler;
}
